docs/manual/index-checker.tex

```diff
@@ -457,8 +460,65 @@ \section{Binary search indices\label{index-searchindex}}
   This is the top type, and programmers should rarely need to write it.
 \end{description}
 
+\section{IndexOf indices\label{index-indexof}}
+
+The sole purpose of the IndexOf Checker is to allow expressing the
+properties of the return value of the JDK's
+\sunjavadoc{java/lang/String.html\#indexOf-java.lang.String-}{String.indexOf} and
+\sunjavadoc{java/lang/String.html\#lastIndexOf-java.lang.String-}{String.lastIndexOf} methods.
+These methods return an index of a substring, or -1 if the substring is not found.
+
+Because of the value -1, the return type of these methods cannot be
+annotated by a \<@LTLengthOf> with a positive offset.
+The \<@IndexOfIndexFor> annotation solves it by having the same
+meaning as \<@LTLengthOf> with the same parameters, except additionally
+allowing the -1 value. The Upper Bound Checker can then infer
+the corresponding \<@LTLengthOf> annotation for expressions that
+are known to be non-negative.
+
+The \<@IndexOfIndexFor(value="this", offset="\#1.length()-1")> annotation on the return value of
+\<String.indexOf> allows writing code such this without warnings from the Upper Bound Checker.
```

Comment:
What would go wrong otherwise?  It usually isn't helpful to readers to write a code snippet and expect the reader to understand the behavior with and without a feature.  (I realize this is clear to you and that it concisely motivates the feature, but you have more context and understanding of the feature than readers do, so it doesn't work for other people.)  I would just cut the example, and as discussed above the whole section.  Organizing the manual in terms of implementation details such as subcheckers is not helpful to a reader.